includes/Core/Util/Feature_Flags.php

```diff
@@ -0,0 +1,140 @@
+<?php
+/**
+ * Class Google\Site_Kit\Core\Util\Feature_Flags
+ *
+ * @package   Google\Site_Kit\Core\Util
+ * @copyright 2020 Google LLC
+ * @license   https://www.apache.org/licenses/LICENSE-2.0 Apache License 2.0
+ * @link      https://sitekit.withgoogle.com
+ */
+
+namespace Google\Site_Kit\Core\Util;
+
+use ArrayAccess;
+
+/**
+ * Class handling plugin activation.
+ *
+ * @since n.e.x.t
+ * @access private
+ * @ignore
+ */
+class Feature_Flags {
+	/**
+	 * Main instance for static access.
+	 *
+	 * @since n.e.x.t
+	 * @var Feature_Flags
+	 */
+	protected static $instance;
+
+	/**
+	 * Feature flag definitions.
+	 *
+	 * @since n.e.x.t
+	 * @var array|ArrayAccess
+	 */
+	private $features;
+
+	/**
+	 * Feature flag mode.
+	 *
+	 * @since n.e.x.t
+	 * @var string
+	 */
+	private $mode;
+
+	/**
+	 * Gets the main feature flags instance.
+	 *
+	 * @since n.e.x.t
+	 * @internal
+	 *
+	 * @return Feature_Flags
+	 */
+	public static function get_instance() {
+		return static::$instance;
+	}
+
+	/**
+	 * Sets the main feature flags instance.
+	 *
+	 * @since n.e.x.t
+	 * @internal
+	 *
+	 * @param Feature_Flags $instance Feature_Flags instance to set as the main instance.
+	 */
+	public static function set_instance( Feature_Flags $instance ) {
+		static::$instance = $instance;
+	}
+
+	/**
+	 * Checks if the given feature is enabled in the current mode on the main instance.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $feature Feature key path to check.
+	 * @return bool
+	 */
+	public static function enabled( $feature ) {
+		return static::get_instance()->is_feature_enabled( $feature );
+	}
+
+	/**
+	 * Constructor.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param array|ArrayAccess $features Feature flag definitions.
+	 * @param string            $mode     Feature flag mode. Default: "production".
+	 */
+	public function __construct( $features, $mode = 'production' ) {
+		$this->features = $features;
+		$this->mode     = $mode;
+	}
+
+	/**
+	 * Checks if the given feature is enabled in the current mode on the main instance.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $feature Feature key path to check.
+	 * @return bool
+	 */
+	public function is_feature_enabled( $feature ) {
+		if ( ! $feature || ! is_string( $feature ) ) {
+			return false;
+		}
+		$feature_path  = explode( '.', $feature );
+		$feature_modes = array_reduce(
+			$feature_path,
+			function ( $value, $key ) {
+				if ( isset( $value[ $key ] ) ) {
+					return $value[ $key ];
+				}
+				return null;
+			},
+			$this->features
+		);
```

Comment:
I think it'd be great to flatten the JSON structure internally once to get an array where the keys are the full identifiers including periods. This will make lookups more efficient because we wouldn't need to flatten on every `is_feature_enabled` call.

Reply:
This doesn't flatten out the entire array of features, it only "drills down" to get the value of the specific feature that was requested so it's only following one "path" through the object. This allows us to support multiple values for `enabled` which is also supported by the webpack plugin that this configuration was based on. Otherwise, we'd have to make certain assumptions about the configuration in order to flatten it (e.g. the value is always assigned to a "enabled" key, even though that isn't required).

If you're concerned about the cost, it would probably be simpler and more efficient to memoize the flattening of `$feature => $value` on-the-fly here rather than flattening the entire structure all at once which would require recursively iterating down all paths (used or not). In reality, I think it's already simple enough that such optimization isn't necessary, but I'm happy to go whichever way you'd like with it 🙂

Reply:
Oh that's a good point actually, I missed that. We shouldn't require people to pass a feature path like `widgets.dashboard.enabled`, the feature path should be `widgets.dashboard`. We can define the convention for how this is declared in the JSON ourselves, so I'm okay assuming that the value is always assigned to an `enabled` key.

We could also get rid of the `enabled` nesting level and put the value right with the actual feature name, but I wouldn't wanna change this now since it would also require changing it everywhere in the JS codebase. Let's alter this code here so that we don't need to pass `enabled`, as that shouldn't be part of the feature name / feature path.